samples/echo/src/main/cpp/AudioEffect.cpp

```diff
@@ -0,0 +1,291 @@
+/*
+ * Copyright 2018 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "AudioEffect.h"
+#include <logging_macros.h>
+#include <climits>
+#include <cstring>
+#include <audio_common.h>
+
+#define CLAMP_FOR_I16(x) \
+  ((x) > SHRT_MAX ? SHRT_MAX : ((x) < SHRT_MIN ? SHRT_MIN : (x)))
+
+/*
+ * Mixing Audio in integer domain to avoid FP calculation
+ *   (FG * ( MixFactor * 16 ) + BG * ( (1.0f-MixFactor) * 16 )) / 16
+ */
+static const int32_t kFloatToIntMapFactor = 128;
+
+AudioMixer::AudioMixer() : AudioFormat(48000, 2, oboe::AudioFormat::I16) {
```

Comment:
Take a look at the files here: https://github.com/google/oboe/tree/master/samples/MegaDrone/src/main/cpp. The mixing and oscillator objects implement a common interface `RenderableAudio`. We should try to put these into a common folder which is shared between the samples. Could be done in a follow up CL though.